### ISSUE TITLE
Yup.Asserts returns interface with all fields marked as required

### ISSUE THREAD

**Issue by macko911** [jquense/yup#1324]
Hi, upgrade to latest minor version `0.32.9` caused some type errors in our codebase.

`Yup.Asserts` returns interface with all fields marked as required, even though they are defined as optional in the schema. I'm wondering if this is desired behaviour or if it's some misunderstanding on my part?

Here's the gist of it, you can also play with it here https://codesandbox.io/s/gracious-leavitt-hjuln?file=/src/index.tsx

```ts
import * as Yup from 'yup';

type Person = {
  id?: string;
}

const schema: Yup.SchemaOf<Person> = Yup.object().shape({
  id: Yup.string(),
});

type Fields = Yup.Asserts<typeof schema>;

// @ts-expect-error Property 'id' is missing in type '{}' but required in type 'AssertsShape<{ id: BaseSchema<string, AnyObject, string>; }>'
const person: Fields = {};
```

Thank you for your feedback!

**Comment by MarcelWaldvogel** [comment]
I just switched to `yup` a few hours ago and I *really, really* enjoyed the `Yup.Asserts<typeof schema>` feature. Wow, thanks!

It seems that the "required" regression was introduced between 0.31.1 and 0.32.0. It would be great if it could be fixed before the 1.0 release.

```typescript
import * as yup from 'yup';

const schema = yup.object({
    a: yup.string(),
    b: yup.object({
        c: yup.string(),
    }).default(undefined),
});

interface S extends yup.Asserts<typeof schema> {}

console.log(schema.validateSync({ a: "foo"}));

function test(s: S) {
  console.log(s);
  console.log(schema.validateSync(s));
}

test({a: "foo"});
test({a: "foo", b: {c: "bar"}});
```

**Comment by MarcelWaldvogel** [comment]
I don't know why it works in 0.31.1 (but it does, even after cleaning `node_modules/` and stuff and doing `npm i` again), as the 0.31.1 code does not define `Asserts` anywhere. Is there an explanation for this?
